internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -51,3 +60,201 @@ func (ar *AlarmsRepository) GetAlarmEventRecordWithUuid(ctx context.Context, uui
+	}
+
+	slog.Info("Successfully inserted alerts from alertmanager", "table", models.AlarmEventRecordView.Name(ctx).String(), "count", len(records))
+	return nil
+}
+
+// buildAlarmEventRecordUpsertQuery builds the query for insert and updating an AlarmEventRecord
+func buildAlarmEventRecordUpsertQuery(records []models.AlarmEventRecord, tableName string) (string, []any, error) {
+	query := psql.Insert(im.Into(tableName))
+
+	// Dynamically construct the column list
+	columns := []string{
+		"alarm_raised_time", "alarm_changed_time", "alarm_cleared_time",
+		"alarm_acknowledged_time", "alarm_acknowledged", "perceived_severity",
+		"extensions", "resource_id", "resource_type_id",
+		"alarm_status", "fingerprint", "alarm_definition_id",
+		"probable_cause_id",
+	}
+
+	// Check if alarm_event_record_id is needed
+	hasAlarmEventRecordID := false
+	for _, record := range records {
+		if record.AlarmEventRecordID != nil {
+			hasAlarmEventRecordID = true
+			columns = append([]string{"alarm_event_record_id"}, columns...)
+			break
+		}
+	}
+
+	// Set the column list for the InsertQuery
+	query.Expression.Columns = columns
+
+	// Add the values for each record
+	values := make([]bob.Mod[*dialect.InsertQuery], 0, len(records))
+	for _, record := range records {
+		// Build the values for the row
+		args := []any{
+			record.AlarmRaisedTime, record.AlarmChangedTime, record.AlarmClearedTime,
+			record.AlarmAcknowledgedTime, record.AlarmAcknowledged, record.PerceivedSeverity,
+			record.Extensions, record.ResourceID, record.ResourceTypeID,
+			record.AlarmStatus, record.Fingerprint, record.AlarmDefinitionID,
+			record.ProbableCauseID,
+		}
+
+		if hasAlarmEventRecordID {
+			if record.AlarmEventRecordID != nil {
+				args = append([]any{record.AlarmEventRecordID}, args...)
+			} else {
+				args = append([]any{nil}, args...)
+			}
+		}
+
+		values = append(values, im.Values(psql.Arg(args...)))
+	}
+
+	// Apply the values to the query
+	query.Apply(values...)
+
+	// constraints only applicable for non-archive table
+	if tableName == models.AlarmEventRecordView.Name(context.Background()).String() {
+		query.Apply(im.OnConflictOnConstraint("unique_fingerprint_alarm_raised_time").DoUpdate(
+			im.SetExcluded("alarm_changed_time"),
+			im.SetExcluded("alarm_cleared_time"),
+			im.SetExcluded("alarm_status"),
+			im.SetExcluded("resource_id"),
+			im.SetExcluded("alarm_definition_id"),
+			im.SetExcluded("probable_cause_id"),
+		))
+	}
+
+	// Compile
+	sql, args, err := query.Build()
+	if err != nil {
+		return sql, args, fmt.Errorf("failed build upsert query for tabls %s: %w", tableName, err)
+	}
+
+	return sql, args, nil
+}
+
+// GetAlarmDefinitions needed to build out aer
+func (ar *AlarmsRepository) GetAlarmDefinitions(ctx context.Context, am *api.AlertmanagerNotification) ([]models.AlarmDefinition, error) {
+	query := psql.Select(
+		sm.Columns("alarm_name", "alarm_definition_id", "probable_cause_id", "resource_type_id"),
```

Comment:
didn't we agree on using the name of the fields from the structure?  We can't have endpoints with different approaches.

Reply:
updated